File: hs/src/Day6.hs
{-|
Module:         Day6
Description:    <https://adventofcode.com/2023/day/6 Day 6: Wait For It>
-}
{-# LANGUAGE OverloadedStrings #-}
module Day6 (part1, part2) where

import Common (readEntire)
import Data.Char (isDigit)
import Data.Function (on)
import Data.List (unfoldr)
import Data.Text (Text)
import qualified Data.Text as T (breakOn, dropWhile, filter)
import qualified Data.Text.Read as T (decimal)

winCount :: Int -> Int -> Int
winCount time distance = ceiling (b + d - 1) - floor (b - d + 1) + 1 where
    b = fromIntegral time / 2
    d = sqrt $ b * b - fromIntegral distance

part1 :: Text -> Int
part1 = product . uncurry (zipWith winCount `on` unfoldr reader) . T.breakOn "\n"
  where reader = either (const Nothing) Just . T.decimal . T.dropWhile (not . isDigit)

part2 :: Text -> Either String Int
part2 input = winCount <$>
    readEntire T.decimal (T.filter isDigit line1) <*>
    readEntire T.decimal (T.filter isDigit line2)
  where (line1, line2) = T.breakOn "\n" input

File: py/aoc2023/day6.py
"""
Day 6: Wait For It
"""

import re
from math import ceil, floor, prod, sqrt

SAMPLE_INPUT = """
Time:      7  15   30
Distance:  9  40  200
"""

NUMBER = re.compile(r"\d+")


def _wincount(time, distance):
    b = time / 2
    d = sqrt(b * b - distance)
    return ceil(b + d - 1) - floor(b - d + 1) + 1


def part1(data):
    """
    >>> part1(SAMPLE_INPUT)
    288
    """
    line1, line2 = filter(None, data.splitlines())
    return prod(
        _wincount(int(time.group()), int(distance.group()))
        for time, distance in zip(NUMBER.finditer(line1), NUMBER.finditer(line2))
    )


def part2(data):
    """
    >>> part2(SAMPLE_INPUT)
    71503
    """
    line1, line2 = filter(None, data.splitlines())
    return _wincount(
        int("".join(NUMBER.findall(line1))), int("".join(NUMBER.findall(line2)))
    )


parts = (part1, part2)
